library/extractor/src/main/java/com/google/android/exoplayer2/metadata/flac/VorbisComment.java

```diff
@@ -18,12 +18,12 @@
 import static com.google.android.exoplayer2.util.Util.castNonNull;
 
 import android.os.Parcel;
-import android.os.Parcelable;
 import androidx.annotation.Nullable;
 import com.google.android.exoplayer2.MediaMetadata;
 import com.google.android.exoplayer2.metadata.Metadata;
 
-/** A vorbis comment. */
+/** @deprecated Use {@link com.google.android.exoplayer2.metadata.vorbis.VorbisComment} instead. */
```

Comment:
Perhaps we can create a package-private supertype called `VorbisCommentImpl` or similar and have the new and deprecated types implement that. I think that's a better approach that preserves the `final` trait and ensures backwards compat.

Reply:
If i'm understanding your proposal correctly, I don't think it does ensure backwards compat for the use-case I described above.

I think you're proposing something like this:

```java
class VorbisCommentImpl {
  public final String key;
  public final String value;

  // equals(), hashcode(), toString() etc.
}

// Existing class
@Deprecated
final class flac.VorbisComment extends VorbisCommentImpl {}

// New class
final class vorbis.VorbisComment extends VorbisCommentImpl {}
```

And the producing could would create instances of `vorbis.VorbisComment`?

The problem is that if the consuming developer is checking something like `entry instanceof flac.VorbisComment`, then in order for that check to continue passing the run-time type of `entry` must be `flac.VorbisComment` or a subclass thereof. And in your proposal `flac.VorbisComment` remains `final`, and therefore `entry` must be **exactly** and instance of `flac.VorbisComment`. I think the only way to keep an `instanceof` check like I've shown passing is to make the new class a subclass of the old one.

Reply:
Yeah, I realized that pretty quickly, which is why I crossed out the comment.

Reply:
Pushed commit 0ea8567 to resolve this.

Reply:
I think this inheritance is the wrong way round to solve the backwards compat use-case I described above.

ExoPlayer will now be producing `vorbis.VorbisComment` instances, but a consuming app will still be checking `instanceof flac.VorbisComment`. Because `flac.VorbisComment` is a sub-class of `vorbis.VorbisComment`, this check will return false when passed a `vorbis.VorbisComment` instance.

I think you need to flip it so that `vorbis.VorbisComment` (the new one) extends `flac.VorbisComment` (the old one). And then app code that is checking `instanceof flac.VorbisComment` will continue to work when passed a new `vorbis.VorbisComment` instance.

Reply:
Okay, I can just move the files.

Reply:
Resolved with f6bee30.